### ISSUE TITLE
Client Crashes When Syncing 300,000+ Files

### ISSUE THREAD

**Issue by elsiehupp** [nextcloud/desktop#3282]
## How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments. 

## Expected Behavior
Nextcloud Desktop should be able to upload a folder with thousands of subfolders and hundreds of thousands of files.

## Actual behaviour
While scanning the folder on disk, Nextcloud Desktop will create a log file hundreds of megabytes in size ([the log file with the date and time in the filename](https://github.com/nextcloud/desktop/blob/4efe00fecaf88b587c6bcb75f1af17ba1fa12747/src/libsync/logger.cpp#L297), not one of the log files named after individual folders); if the Settings window is in the foreground, the cursor will beachball; and ultimately Nextcloud Desktop will close unexpectedly.

Considering the log file will continue to grow (and I've seen it exceed 750MB), it's possible that the crash is being caused by the part of the application that is writing the logs. This may qualify as a "memory leak", and indeed there is a very old Issue regarding excessive memory usage by the client: https://github.com/nextcloud/desktop/issues/124. Also https://github.com/nextcloud/desktop/issues/1422 seems to deal with excessive memory usage.

If indeed the logging is the problem, perhaps the log buffer could have a cap on the number of lines, or if a log file exceeds a certain number of lines, the logger could automatically start a new file to prevent the buffer from taking up too much space in memory, idk.

## Steps to reproduce
1. Try to sync a folder with thousands of subfolders and hundreds of thousands of files.

## Client configuration
Client version: 3.2.1 (though this behavior is not new)
Operating system: macOS 11 Big Sur
OS language: English
Installation path of client: `/Applications`

## Logs
The log file appears to be the bug. That is, I'm not going to upload a 750MB log file here.

**Comment by allexzander** [comment]
@elsiehupp Will the files get synchronized eventually if reopening the desktop client after it crashes?

**Comment by elsiehupp** [comment]
> @elsiehupp Will the files get synchronized eventually if reopening the desktop client after it crashes?

I actually didn’t try repeatedly reopening the application in order to see if it eventually succeeds. I was feeling a bit low on patience, so I just excluded the problem directory from my sync.

The most immediate problem here is that the application is repeatedly crashing, so it doesn’t complete a sync regardless. The user shouldn’t have to babysit the sync operation (though there are indeed several other things that require some degree of babysitting, but those are separate issues).

Basically, the application should handle overwhelming volumes of data gracefully and restart a failed sync without user intervention. (And, more importantly, each time the sync automatically restarts, it should pick up where it left off, so that it actually makes incremental progress over time.)

If you think it would be useful, I could try and repeatedly manually restart the sync with the problem directory, but personally I think the core issue of the application drowning in its own logs is pretty easy to diagnose from the existing information.

**Comment by elsiehupp** [comment]
I just added a 60MB logfile that IIRC didn’t result in a crash.

**Comment by elsiehupp** [comment]
It seems to be doing something similar on my Linux computer, too. The logs folder swelled to over 12GB, with individual log files up to 500MB each:

<img width="1024" alt="Screen Shot 2021-06-21 at 10 38 47 PM" src="https://user-images.githubusercontent.com/9206310/123545055-3d2bb080-d724-11eb-93b7-c864004dd51d.png">

<img width="1024" alt="Screen Shot 2021-06-21 at 10 39 32 PM" src="https://user-images.githubusercontent.com/9206310/123545063-474daf00-d724-11eb-89d5-6f0630e997c0.png">

**Comment by elsiehupp** [comment]
It looks like the maasive-log-file problem afflicts Nextcloud Server, too (though obviously they have a separate issue tracker):

<img width="433" alt="Screen Shot 2021-08-09 at 2 41 07 AM" src="https://user-images.githubusercontent.com/9206310/128788444-4ff38806-e6f3-4f58-8dc1-ffb33d9011d7.png">
<img width="437" alt="Screen Shot 2021-08-09 at 2 41 44 AM" src="https://user-images.githubusercontent.com/9206310/128788451-a6b526cd-6316-47da-8fe6-14e033075ab1.png">
<img width="528" alt="Screen Shot 2021-08-09 at 2 42 22 AM" src="https://user-images.githubusercontent.com/9206310/128788460-3bddf083-b252-4d04-8423-7bfa4e2d69b1.png">

**Comment by tele81** [comment]
whats the solution?

**Comment by arch-user-france1** [comment]
> @elsiehupp Will the files get synchronized eventually if reopening the desktop client after it crashes?

probably not, because it takes up to 99% CPU usage and does not respond to user interaction.

**Comment by Reiikz** [comment]
+1 on this.
WTF is going on, now I've collected the logs untill it reached 16GB of memory usage then killed it.
No amount of syncing files with rsync -a or reopeining and even starting the entire download from the server on a clean directory fixes this.
It eventually breaks, the UI becomes more unresposive than it generally is and eventually it starts leaking memory to the point it would fill any reasonable person's ram.
I happen to have 128GB of ram and it filled mine!
This needs urgent fix.
I can provide logs but I need assistance anonimizing them as they have hundreds of thousands of paths and checksums of private files.

**Comment by Rello** [comment]
Hello,

this ticket is pretty old and the client had several fundamental redesigns.
If the issue still exists, please create a new ticket with current information.
100k+ files is a common case for customers. the syncing was optimised during recent releases, so was the logging.

Thank you
